src/main/java/frc/robot/subsystems/drive/Drive.java

```diff
@@ -0,0 +1,437 @@
+        // TODO either write custom handler for gyro angle or beg for pigeon 2
+        // vision also corrects for async odo so maybe just cope
+        double[] sampleTimestamps =
+            modules[0].getOdometryTimestamps(); // All signals are sampled together
+        int minTimestamps =
+            min(
+                min(
+                    modules[0].getOdometryTimestamps().length,
+                    modules[1].getOdometryTimestamps().length),
+                min(
+                    modules[2].getOdometryTimestamps().length,
+                    modules[3].getOdometryTimestamps().length)); // Check if one wheel data was rejected
+        int sampleCount;
+        if (gyroInputs.connected) {
+            sampleCount = min(minTimestamps, gyroInputs.odometryYawPositions.length);
+        } else {
+            sampleCount = minTimestamps;
+        }
+        for (int i = 0; i < sampleCount; i++) {
+            // Read wheel positions and deltas from each module
+            SwerveModulePosition[] modulePositions = new SwerveModulePosition[4];
+            SwerveModulePosition[] moduleDeltas = new SwerveModulePosition[4];
+            for (int moduleIndex = 0; moduleIndex < 4; moduleIndex++) {
+                modulePositions[moduleIndex] = modules[moduleIndex].getOdometryPositions()[i];
+                moduleDeltas[moduleIndex] =
+                    new SwerveModulePosition(
+                        modulePositions[moduleIndex].distanceMeters
+                            - lastModulePositions[moduleIndex].distanceMeters,
+                        modulePositions[moduleIndex].angle);
+                lastModulePositions[moduleIndex] = modulePositions[moduleIndex];
+            }
+
+            // Update gyro angle
+            if (gyroInputs.connected) {
+                // Use the real gyro angle
+                rawGyroRotation = gyroInputs.odometryYawPositions[i];
+            } else {
+                // Use the angle delta from the kinematics and module deltas
+                Twist2d twist = kinematics.toTwist2d(moduleDeltas);
+                rawGyroRotation = rawGyroRotation.plus(new Rotation2d(twist.dtheta));
+            }
+
+            // Apply update
+            poseEstimator.updateWithTime(sampleTimestamps[i], rawGyroRotation, modulePositions);
+        }
+        Logger.recordOutput(
+            "Drive/Running Command",
+            Optional.ofNullable(this.getCurrentCommand()).map(Command::getName).orElse("None"));
+    }
+
+    /**
+    * Runs the drive at the desired velocity.
+    *
+    * @param speeds Speeds in meters/sec
+    */
+    public void runVelocity(ChassisSpeeds speeds) {
+        // Calculate module setpoints
+        ChassisSpeeds discreteSpeeds = ChassisSpeeds.discretize(speeds, 0.02);
+        SwerveModuleState[] setpointStates = kinematics.toSwerveModuleStates(discreteSpeeds);
+        SwerveDriveKinematics.desaturateWheelSpeeds(setpointStates, DriveConstants.MAX_LINEAR_SPEED);
```

Comment:
I would suggest a before and after on this desaturation call in case this becomes relevant later.

Reply:
To clarify, ideally there should be no desaturation needed (we are always commanding the wheels to go no faster than they are capable). Any desaturation occurring is error that the caller of runVelocity now is surprised about and has to deal with on its feedback loop (PID) as opposed to the feedforward loop (trapezoidal profile, among other things). Does this make sense, and do you agree?

I don't know if it will happen in practice, but clearly Ishaan put it there for a reason, probably because it was happening. Might be worth an ask to Ishaan to see why this is there, and assuming there isn't a good answer, raise some kind of logger message or clear error that this is not expected.

This is not important enough to block the review though.